examples/step-67/step-67.cc

```diff
@@ -1465,60 +1505,362 @@ namespace Euler_DG
+
+                if (factor_ai != Number())
+                  phi_temp.reinit(cell);
+
+                if (factor_ai != Number() && stage == 0)
+                  {
+                    phi.read_dof_values(src);
+
+                    for (unsigned int i = 0;
+                         i < phi.static_dofs_per_component * (dim + 2);
+                         ++i)
+                      phi_temp.begin_dof_values()[i] =
+                        phi.begin_dof_values()[i];
+
+                    phi.evaluate(EvaluationFlags::values);
+                  }
+                else
+                  {
+                    phi.gather_evaluate(src, EvaluationFlags::values);
+                  }
+
+                for (unsigned int i = 0; i < phi.static_n_q_points * (dim + 2);
+                     ++i)
+                  buffer[i] = phi.begin_values()[i];
+
+                for (unsigned int q = 0; q < phi.n_q_points; ++q)
+                  {
+                    const auto w_q = phi.get_value(q);
+                    phi.submit_gradient(euler_flux<dim>(w_q), q);
+                    if (body_force.get() != nullptr)
+                      {
+                        const Tensor<1, dim, VectorizedArray<Number>> force =
+                          constant_function ?
+                            constant_body_force :
+                            evaluate_function<dim, Number, dim>(
+                              *body_force, phi.quadrature_point(q));
+
+                        Tensor<1, dim + 2, VectorizedArray<Number>> forcing;
+                        for (unsigned int d = 0; d < dim; ++d)
+                          forcing[d + 1] = w_q[0] * force[d];
+                        for (unsigned int d = 0; d < dim; ++d)
+                          forcing[dim + 1] += force[d] * w_q[d + 1];
+
+                        phi.submit_value(forcing, q);
+                      }
+                  }
+
+                if (body_force.get() == nullptr)
+                  {
+                    for (unsigned int i = 0;
+                         i < phi.static_n_q_points * (dim + 2);
+                         ++i)
+                      phi.begin_values()[i] = 0.0;
+                  }
+
+                VectorizedArray<Number> *values_ptr  = phi.begin_values();
+                VectorizedArray<Number> *grdient_ptr = phi.begin_gradients();
+
+                for (unsigned int c = 0; c < dim + 2; ++c)
+                  {
+                    if (dim >= 1)
+                      eval.template gradients<0, false, true>(
+                        grdient_ptr + phi.static_n_q_points * 0, values_ptr);
+                    if (dim >= 2)
+                      eval.template gradients<1, false, true>(
+                        grdient_ptr + phi.static_n_q_points * 1, values_ptr);
+                    if (dim >= 3)
+                      eval.template gradients<2, false, true>(
+                        grdient_ptr + phi.static_n_q_points * 2, values_ptr);
```

Comment:
Should we look into possibilities to hide some of these functions behind library functions?